src/demikernel/libos/network/queue.rs

```diff
@@ -247,67 +214,56 @@ impl<T: NetworkTransport> SharedNetworkQueue<T> {
     where
         F: FnOnce() -> Result<QToken, Fail>,
     {
-        self.state_machine.prepare(SocketOp::Close)?;
-        self.do_generic_sync_control_path_call(coroutine_constructor)
+        self.state_machine.ensure_not_closing()?;
+        coroutine_constructor()
     }
 
     /// Close this queue. This function contains all the single-queue functionality to synchronously close a queue.
     pub fn hard_close(&mut self) -> Result<(), Fail> {
-        //self.state_machine.prepare(SocketOp::Close)?;
-        //self.state_machine.commit();
-        match self.transport.clone().hard_close(&mut self.socket) {
-            Ok(()) => {
-                //self.state_machine.prepare(SocketOp::Closed)?;
-                //self.state_machine.commit();
-                Ok(())
-            },
-            Err(e) => Err(e),
-        }
+        self.state_machine.ensure_not_closing()?;
+        self.transport.clone().hard_close(&mut self.socket)?;
+        self.state_machine.closed();
+        Ok(())
     }
 
     /// Asynchronously closes this queue. This function contains all of the single-queue, asynchronous code necessary
     /// to close a queue and any single-queue functionality after the close completes.
     pub async fn close_coroutine(&mut self) -> Result<(), Fail> {
-        match self.transport.clone().close(&mut self.socket).await {
-            Ok(()) => {
-                self.state_machine.prepare(SocketOp::Closed)?;
-                self.state_machine.commit();
-                Ok(())
-            },
-            Err(e) => Err(e),
+        self.state_machine.closing();
+        {
```

Comment:
Is this artificial scoping intended to express clarity/separation of responsibilities, or for scoping the variables inside? If it's intended for clarity, the inner block can be a separate well-named fn instead. That ways, the reader can easily grasp what's happening.

Reply:
It is for scoping ownership for the socket. I wanted to avoid cloning if not necessary, but I will use a separate function.